lib/Elevate/Components/Grub2ControlTest.pm

```diff
@@ -0,0 +1,200 @@
+package Elevate::Components::Grub2ControlTest;
+
+=encoding utf-8
+
+=head1 NAME
+
+Elevate::Components::Grub2ControlTest
+
+=head2 mark_cmdline
+
+Add a random option to the grub cmdline
+
+=head2 verify_cmdline
+
+Verify that the option is present after a reboot and block if it is not
+
+=cut
+
+use cPstrict;
+
+use Elevate::OS        ();
+use Elevate::StageFile ();
+
+use Log::Log4perl qw(:easy);
+use File::Slurper ();
+
+use parent qw{Elevate::Components::Base};
+
+use constant GRUBBY_PATH                  => '/usr/sbin/grubby';
+use constant UPDATE_GRUB_PATH             => '/usr/sbin/update-grub';                            # This is an official alias for grub-mkconfig with the correct arguments
+use constant GRUB_MKCONFIG_FRAG_DIR_PATH  => '/etc/default/grub.d';
+use constant GRUB_MKCONFIG_FRAG_FILE_PATH => GRUB_MKCONFIG_FRAG_DIR_PATH . '/zzz-elevate.cfg';
+use constant CMDLINE_PATH                 => '/proc/cmdline';
+
+# In place of Unix::Sysexits:
+use constant EX_UNAVAILABLE => 69;
+
+sub _call_grubby ( $self, @args ) {
+
+    my %opts = (
+        should_capture_output => 0,
+        should_hide_output    => 0,
+        die_on_error          => 0,
+    );
+
+    if ( ref $args[0] eq 'HASH' ) {
+        my %opt_args = %{ shift @args };
+        foreach my $key ( keys %opt_args ) {
+            $opts{$key} = $opt_args{$key};
+        }
+    }
+
+    unshift @args, GRUBBY_PATH;
+
+    return $opts{die_on_error} ? $self->ssystem_and_die(@args) : $self->ssystem( \%opts, @args );
+}
+
+sub _call_update_grub ( $self, $opts = {} ) {
+    die "Argument must be a hashref" unless ref $opts eq 'HASH';
+    return $opts->{die_on_error} ? $self->ssystem_and_die(UPDATE_GRUB_PATH) : $self->ssystem(UPDATE_GRUB_PATH);
+}
+
+sub _default_kernel ($self) {
+    return $self->_call_grubby( { should_capture_output => 1, should_hide_output => 1 }, '--default-kernel' )->{'stdout'}->[0] // '';
+}
+
+sub _persistent_id {
+    my $id = Elevate::StageFile::read_stage_file( 'bootloader_random_tag', '' );
+    return $id if $id;
+
+    $id = int( rand(100000) );
+    Elevate::StageFile::update_stage_file( { 'bootloader_random_tag', $id } );
+    return $id;
+}
+
+sub _add_kernel_arg ( $self, $arg ) {
+
+    if ( Elevate::OS::bootloader_config_method() eq 'grubby' ) {
```

Comment:
It can be argued that this is a violation of the `Cpanel::OS` principles which are (probably) carried over into `Elevate::OS`. The counterargument is that the correct way to handle this, which is to create a submodule for each set of behaviors, sounds like a lot of boilerplate with not so much meat.